File: trello-app/src/service/card.service.ts
import { Injectable } from '@angular/core';
import { AuthService } from './auth.service';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Card } from 'src/model/card';

@Injectable({
  providedIn: 'root'
})
export class CardService {
  readonly BASE_URL = 'http://localhost:8098/trello';

  constructor(private httpClient: HttpClient, private authService: AuthService) {
  }

  create(title: string, boardId: string, cardListId: string): Observable<Card> {
    return this.httpClient.post<Card>(`${this.BASE_URL}/boards/${boardId}/cardlists/${cardListId}/cards`, title, {
      headers: this.authService.getAuthHeader()
    });
  }

  update(boardId: string, cardListId: string, cardId: string, card: Card): Observable<Card> {
    return this.httpClient.put<Card>(`${this.BASE_URL}/boards/${boardId}/cardlists/${cardListId}/cards/${cardId}`, card, {
      headers: this.authService.getAuthHeader()
    });
  }

  move(sourceCardListId: string, previousIndex: number,
       targetCardListId: string, currentIndex: number, boardId: string): Observable<{}> {
    return this.httpClient
        .put(`${this.BASE_URL}/boards/${boardId}/cardlists/${sourceCardListId}/cards/move`, {}, {
      headers: this.authService.getAuthHeader(),
      params: new HttpParams()
        .append('targetCardListId', targetCardListId)
        .append('previousIndex', previousIndex.toString())
        .append('currentIndex', currentIndex.toString())
    });
  }
}

File: trello-server/src/main/java/pl/trello/service/CardService.java
package pl.trello.service;

import org.springframework.stereotype.Service;
import pl.trello.core.NotFoundException;
import pl.trello.core.OrderedUtils;
import pl.trello.core.Utils;
import pl.trello.model.Card;
import pl.trello.model.CardList;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class CardService {

    private CardListService cardListService;

    public CardService(CardListService cardListService) {
        this.cardListService = cardListService;
    }

    public Card getById(String boardId, String cardListId, String cardId) throws NotFoundException {
        return cardListService.getById(boardId, cardListId).getCards().stream()
                .filter(card -> card.getId().equals(cardId)).findAny()
                .orElseThrow(() -> new NotFoundException("Card not found"));
    }

    public Card create(String boardId, String cardListId, String title) throws NotFoundException {
        CardList cardList = cardListService.getById(boardId, cardListId);
        Card card = build(title, cardList);
        cardList.getCards().add(card);
        cardListService.update(boardId, cardList);
        return card;
    }

    public void move(String boardId, String sourceCardListId, int previousIndex,
                     String targetCardListId, int currentIndex) throws NotFoundException {
        CardList source = cardListService.getById(boardId, sourceCardListId);
        if (!sourceCardListId.equals(targetCardListId)) {
            CardList target = cardListService.getById(boardId, targetCardListId);
            OrderedUtils.move(source.getCards(), previousIndex, target.getCards(), currentIndex);
            cardListService.update(boardId, target);
        } else {
            OrderedUtils.move(source.getCards(), previousIndex, currentIndex);
        }
        cardListService.update(boardId, source);
    }

    public Card update(String boardId, String cardListId, String cardId, Card cardUpdate) throws NotFoundException {
        getById(boardId, cardListId, cardId);
        CardList cardList = cardListService.getById(boardId, cardListId);
        cardList.setCards(updateCards(cardList.getCards(), cardUpdate));
        cardListService.update(boardId, cardList);
        return cardUpdate;
    }

    private List<Card> updateCards(List<Card> cards, Card cardUpdate) {
        return cards.stream()
                .map(card -> card.getId().equals(cardUpdate.getId()) ? cardUpdate : card)
                .collect(Collectors.toList());
    }

    private Card build(String title, CardList cardList) {
        return Card.builder()
                .id(Utils.generateId())
                .title(title)
                .description("")
                .order(cardList.getCards().size())
                .checklists(new ArrayList<>())
                .comments(new ArrayList<>())
                .build();
    }
}
